### PR TITLE
ar71xx: add support for TL-WR1043N v5.0

### PR DESCRIPTION
**Pull request by majik-sheff**
The hardware of the TP-Link WR-1043N v5 appears
to be to almost identical to the WR-1043NDv4, except
that the USB port has been removed and there is no
longer a removeable antenna option.

The software is more in line with the Archer series in
that is uses a nested bootloader scheme. The factory
image should install from the stock interface.

Signed-off-by: Tim Thorpe <tim@tfthorpe.net>

### CODE REVIEW

**Comment by Brother-Lal** [comment]
Hi!
Very nice that there is an effort for the support. Good work!

You said that it is identical to the 1043v4 - could you just take the file for the v4 and add a definition for the v5 there? So that there is not a complete copy of the settings in the v5 machfile?

Or are there to many differences?

Greetings,

Brother_Lal

**Comment by majik-sheff** [comment]
I submitted this PR as it was because I finally had it in a usable state.  LEDs work, buttons work, WiFi works, Ethernet works, etc etc.  

I was a bit hasty in my description I guess.  When I said Identical, I meant in terms of hardware specs and and device specifics (sans USB of course).  The software side uses their more recent safeloader setup which uses nested bootloaders and partition tables.  It's like a hybrid between the WR-1043 v4 and an Archer.

Having said that, the only things that I am aware of that are actually broken are related to installation.  I have a full dump of the original firmware and am trying to pick apart a way to convince the factory firmware to take the image.  Sysupgrade will work after the initial TFTP install, but only if you use the -F option.  It's my understanding that sysupgrade images should have a special header to denote their nature.  I am really struggling to fully understand the image build system and would appreciate some input from some other contributors who have worked on getting support for the Archer series.

You are right that I can probably implement the v5 main file as a variant of the v4.  The main issue is that I built my file primarily off of the Dragonfly reference and the GPL provided file from TP-Link.  Now that I'm looking at the files side-by-side I can see that it probably won't be an issue (famous last words) to make a single file out of them..  I'll just have to sit down with it and figure out how to go about it.

**Comment by Brother-Lal** [comment]
I usually take the "lazy bastard" route and just copy a file that has the most similarities or add it in the same ...
I don't know how the policy is at what stage you can use the same file, but i have the impression even the GPIOs are alike, so there  could be a lot of reuse.

Some comments at the PR:

I am missing a "firmware" partition in the MTDPARTS - i think it is necessary for the flash to work, and should encompass kernel+rootfs starting at the first partition of the two (kernel) - that could be your sysupgrade problem?

Regarding the firmware, the ArcherC25 had to have a hack (https://github.com/lede-project/source/commit/9cd69c45c505b5f81b8fa78002130a635e604bef#diff-b232b41a358a2b44c64fb32925bd265aR1045 ) to accept the image, because the "extra-para" partitioton is needed.

**Comment by majik-sheff** [comment]
I bet you're right about the sysupgrade bit. I'll look the hack when I have some time.

File: target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr1043n-v5.c [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+		.active_low	= 1,
+	},
+	{
+		.name		= "tp-link:green:system",
+		.gpio		= TL_WR1043N_V5_GPIO_LED_STATUS,
+		.active_low	= 1,
+		.default_state = LEDS_GPIO_DEFSTATE_ON,
+	},
+	{
+		.name		= "tp-link:green:lan_1",
```
Comment (f00b4r0) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr1043n-v5.c]: why '_1'? It seems most of the other board use lan1, lan2, etc. Might be a good idea to stick to the same naming conventions, especially across similar boards.

File: target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr1043n-v5.c [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+		.gpio		= TL_WR1043N_V5_GPIO_LED_LAN3,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tp-link:green:lan_4",
+		.gpio		= TL_WR1043N_V5_GPIO_LED_LAN4,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tp-link:orange:wan_link",
```
Comment (f00b4r0) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr1043n-v5.c]: same here, why not just 'wan'?

File: target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr1043n-v5.c [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+static struct mdio_board_info tl_wr1043n_v5_mdio0_info[] = {
+	{
+		.bus_id = "ag71xx-mdio.0",
+		.phy_addr = 0,
+		.platform_data = &tl_wr1043n_v5_ar8337_data,
+	},
+};
+
+static void __init tl_wr1043nv5_setup(void)
```
Comment (f00b4r0) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr1043n-v5.c]: several indentation problems in that function, please review.

Comment (majik-sheff) [reply target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr1043n-v5.c]: I admit it was a splice job between the v4 file and the TP-Link GPL sources.  I hope to have time to clean up the code in the next few days.  Thank you for notes.

**Comment by Brother-Lal** [comment]
I do have a v5 now residing here! I will try to test your PR and have a look at the original firmware files + sysupgrade.

**Comment by Brother-Lal** [comment]
Ok, the firmware is now accepted from the webinterface with these changes:
https://github.com/Brother-Lal/source/commit/38cd8c0d672b6718ebdae0f3596b165b1cfb15c9

File: target/linux/ar71xx/base-files/etc/board.d/01_leds [hunk 4]
```diff
@@ -799,6 +799,14 @@ tl-wr1043nd-v4)
 	ucidef_set_led_switch "lan3" "LAN3" "tp-link:green:lan3" "switch0" "0x04"
 	ucidef_set_led_switch "lan4" "LAN4" "tp-link:green:lan4" "switch0" "0x02"
 	;;
+tl-wr1043n-v5)
```
Comment (Brother-Lal) [inline target/linux/ar71xx/base-files/etc/board.d/01_leds]: If we want to merge v4 and v5, we can merge here too, just add the usb for v4 only

**Comment by majik-sheff** [comment]
Yes.  That was on my short list of things to fix.

**Comment by Brother-Lal** [comment]
ok, with my last change [here](https://github.com/Brother-Lal/source/commit/358faa593b7739cff35780b0f23456812e0df28e) sysupgrade works (and remember to try the sysupgrade image, not the factory image for sysupgrade, it saves testing time looking for checkimage failed :/ )
For me this looks as if feature-wise everything works 👍 

Final question, merging into 1043v4: do you want to do that, shall I, or has anybody else an opinion if its worthy/necessary? 😬

**Comment by Brother-Lal** [comment]
Ok, i got a merged version with the wifi mac set properly in my branch:
https://github.com/Brother-Lal/source/tree/1043v5

@majik-sheff can you update your PR?

Oh, and the "signed-of-by" should contain the real name, as far as i know.

**Comment by majik-sheff** [comment]
I think I updated it to properly reflect the current state.

**Comment by rotanid** [comment]
hm, can't see any update to this PR - maybe you mixed something up?

**Comment by Brother-Lal** [comment]
Rewriting git history can be a pain at some times!

For me, I did a "git rebase -i HEAD~5" on my commits, then chose to "r"eword your first commit and "s"quash the other ones - this fixes the commit itself + merges all of my fixes into your commit for a clean integration into the main repo :)

Should look like this then:
https://github.com/Brother-Lal/source/commit/db5dfc4e7bbdba5881acb5e9e97e36f20222e71c

Note the commit message with the same convention as the PR, and the signed of lines :)

**Comment by Brother-Lal** [comment]
@majik-sheff Any chances that you can rework the PR this week, or should I submit an own?
We do have some 1043v5 in stock now waiting for software, and a backport to 17.01 would be needed for them also, which i would want to base on the commited support in mainline :)

**Comment by majik-sheff** [comment]
Sorry, it's been a hectic fall :)  I have some time to work on it today, so that's what I'm doing. Gotta read through the notes again and sync my local repo.  I'll try to get an updated PR together.  Thanks for sticking with me, this has certainly been an education.

**Comment by majik-sheff** [comment]
I believe I have successfully updated my branch and squashed the commit into the PR.  The mach-tl-wr1043nd-v4.c file looks like a hot mess because the indentation styles were all over the place and I just forced the formatting to be consistent.  If it goes against LEDE coding convention, I apologize.  Thank you again for all of your help in getting this to a usable state.  I'm sure my next contribution will go much more smoothly.  :)

**Comment by Brother-Lal** [comment]
Nice job integrating all that!

The formatting changes may be meant good, but i think they might bite you back:
Do a:
> git format-patch --stdout -1 | ./scripts/checkpatch.pl - 
to check if the patch follows the coding guidelines...

Can you revert them and then run the checkscript only on your new changes (which it does automagically?)
Some errors may be ignorable (i would leave the safeloader if in one line, until more are added, for example, in my opinion)

And yes, as soon as you know the steps that are necessary to align a patch with the standards, its only the comments from the people with commit rights which will be left over :D 

PS: Commit message has an "a" missing, and i would shorten it a bit - most of this info should be put into the wiki :)

**Comment by majik-sheff** [comment]
I may have found an issue with the factory image.  For some reason the firmware partition wasn't created during the upgrade.  I have a stack of fresh ones here to test on.  When I can be sure it wasn't a bad build I'll fix up those last few things.

**Comment by adridolf** [comment]
Nice work. In case someone is interested, I had been able to rebase this to lede-17.01 with the following adjustments:
- Apply: https://github.com/majik-sheff/source/commit/b3cb0e75884222e141a40443c25d20deedf7a552
- Apply: https://github.com/majik-sheff/source/commit/f08f754993d5e9a1952400528ce5405910d1d1a4
- Change TPLINK_BOARD_ID back to TPLINK_BOARD_NAME (to get around https://github.com/majik-sheff/source/commit/7d6c63d875421fc1f66833ba0e690914ea28f1a3)
- Playing around with the correct positions in tplink.mk, since this patch is missing in lede-17.01: https://github.com/majik-sheff/source/commit/7efe0cd83d343d67f69c705844ca468a764a939c

Working like a charm in combination with a Freifunk firmware, flashing through the original firmware's WebUI was successful.
For other devices than the 1043v5, additional adjustments may be required.

**Comment by majik-sheff** [comment]
Wow!  That really helps to clean up a lot of the messy hacks.  Looking at it now I think they were half of the reason I struggled to understand how to bring the v5 in.

**Comment by adridolf** [comment]
Although this is unrelated to the PR itself, may someone can help:
I need two mac-addresses for the device, the one on the outside I get without problems, but with the second one I have problems. For the 1043v4, one could do:

mtd_get_mac_binary config 0x1017c

Is there a similar possibility for the 1043v5?

**Comment by majik-sheff** [comment]
As far as I can tell, there is no second MAC address stored in the config partition.  All other MACs are computed.  You can manually set the address either in the advanced section of interfaces in luci or by editing /etc/config/network to include:

option macaddr 'xx:xx:xx:xx:xx:xx'

in the interface section

**Comment by rotanid** [comment]
@majik-sheff could you please take care of the changes @Brother-Lal  asked for so that this PR can go forward? i really like to have this device integrated soon!

**Comment by rotanid** [comment]
as we really need this support and there is no answer here, @Brother-Lal opened a new pull request for this at #1506

**Comment by majik-sheff** [comment]
Sorry.  Real life got in the way of me being able to sit down with this.  If someone else hasn't already taken it, I can fix it up today.

**Comment by Brother-Lal** [comment]
No problems - real life comes first.

I already submitted a new PR(https://github.com/lede-project/source/pull/1506) and will fix any issues until it's merged, so this one can be closed :)
Thanks anyway for your contributions!

**Comment by majik-sheff** [comment]
OK, thank you.  I really appreciate you picking that up.  I hated to leave it so close to finished.  This is a fantastic community.

**Comment by dedeckeh** [comment]
Closing PR as review is ongoing in https://github.com/lede-project/source/pull/1506

**Pull request closed without merging**